packages/react-dom/src/shared/trustedTypesAwareToString.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+/**
+ * Returns true only if Trusted Types are available in global object and the value
+ * is a trusted type.
+ */
+function isTrustedTypesValue(value: any) {
+  if (typeof window.TrustedTypes === 'undefined') {
+    return false;
+  } else {
+    return (
+      window.TrustedTypes.isHTML(value) ||
+      window.TrustedTypes.isScript(value) ||
+      window.TrustedTypes.isScriptURL(value) ||
+      window.TrustedTypes.isURL(value)
+    );
+  }
+}
+
+/**
+ * We allow passing objects with toString method as element attributes or in dangerouslySetInnerHTML
+ * and we do validations that the value is safe. Once we do validation we want to use the validated
+ * value instead of the object (because object.toString may return something else on next call).
+ *
+ * If application uses Trusted Types we don't stringify trusted values, but preserve them as objects.
+ */
+function trustedTypesAwareToString(value: any) {
```

Comment:
Please annotate the return value type. I think currently it's typed as `any`, which means all types downwards are wrong. E.g. `toString()` from `ToStringValue` claims to return a string — but after this change, this is no longer guaranteed. We should encode this in types.

Reply:
Yeah, my bad. 

Btw. are you ok with typing the return value as `string`? Be aware that TT are not equivalent to `string` and are missing methods like `.replace`

Reply:
> Be aware that TT are not equivalent to string and are missing methods like .replace

Right — that's what I want to catch with types. So it can be some opaque type as long as we don't consider it a string.

Since Flow probably doesn't support TT at the moment, we can probably have a module with dangerous DOM sinks which are typed to accept our opaque type.